controller/sync.go

```diff
@@ -270,23 +272,43 @@ func (sc *syncContext) sync() {
 	// if it is the last phase/wave and the only remaining tasks are non-hooks, the we are successful
 	// EVEN if those objects subsequently degraded
 	// This handles the common case where neither hooks or waves are used and a sync equates to simply an (asynchronous) kubectl apply of manifests, which succeeds immediately.
-	complete := tasks.Find(func(t *syncTask) bool { return t.phase != phase || wave != t.wave() || t.isHook() }) == nil
+	complete := !tasks.Any(func(t *syncTask) bool { return t.phase != phase || wave != t.wave() || t.isHook() })
 
-	sc.log.WithFields(log.Fields{"phase": phase, "wave": wave, "tasks": tasks}).Debug("filtering tasks in correct phase and wave")
+	sc.log.WithFields(log.Fields{"phase": phase, "wave": wave, "tasks": tasks, "syncFailTasks": syncFailTasks}).Debug("filtering tasks in correct phase and wave")
 	tasks = tasks.Filter(func(t *syncTask) bool { return t.phase == phase && t.wave() == wave })
 
 	sc.setOperationPhase(v1alpha1.OperationRunning, "one or more tasks are running")
 
 	sc.log.WithFields(log.Fields{"tasks": tasks}).Debug("wet-run")
 	if !sc.runTasks(tasks, false) {
-		sc.setOperationPhase(v1alpha1.OperationFailed, "one or more objects failed to apply")
+		sc.runFailedTasksIfAny(syncFailTasks, "one or more objects failed to apply")
 	} else {
 		if complete {
 			sc.setOperationPhase(v1alpha1.OperationSucceeded, "successfully synced (all tasks run)")
 		}
 	}
 }
 
+func (sc *syncContext) runFailedTasksIfAny(syncFailTasks syncTasks, failMessage string) {
+	if len(syncFailTasks) > 0 {
+		// If tasks are already completed, don't run them again
+		if syncFailTasks.All(func(task *syncTask) bool { return task.completed() }) {
```

Comment:
No test coverage?

Reply:
Note - you only get test coverage from tests in the same directory.

Reply:
Added tests in the same directory! https://github.com/argoproj/argo-cd/pull/1795/files#diff-eb441ec1bfbc70512fca4942a3d0915eR493

Reply:
https://github.com/argoproj/argo-cd/blob/54e6d2efcc983ae372a43d0b95b1984feeb24838/controller/sync_test.go#L493-L569